### DISCUSSION TITLE
Navigate menu selection items with `Alt-hjkl` key style

### DISCUSSION THREAD

**Discussion by volodymyrprokopyuk** [marlonrichert/zsh-autocomplete#360] [Q&A] [answered]
## What do you want?

Navigate items in the selection menu using `Alt-h`, `Alt-j`, `Alt-k` and `Alt-l` key combination in the same way the `right`, `down`, `up` and `left` cursor keys work. As I understand this is a matter of `bindkey -M menuselect 'Alt-h' menu-selection-command`, but I do not know 1) how to define `Alt-key` combination for the above keys and 2) what the exact menu selection command should I use (possibly `down-line-or-select` and `up-line-or-search`)

## Why do you want this?

By using the proposed key combinations (along with the existing cursor keys) the left hand does not have to be moved to the cursor keys and so is much faster, comfortable and convenient. A menu selection item can be chosen without moving the left hand from the main position on the keyboard

## Who else would benefit from this?

All the users that use vim and vi-style keybindings in other applications will want and benefit from the proposed enhancement

## Example scenario

Given the following situation:

* After entering the `cd` + `space` command a selection menu of possible directories is presented

When I perform the following steps:

1. When the `Alt+j` key combination is entered

Then I expect the following to happen:

*  The next item in the selection menu is chosen and highlighted (corresponding actions are taken for the rest of the `Alt+hjkl` key combinations`)

**Comment by marlonrichert** [comment, marked as answer]
Menu selection itself is not a feature `zsh-autocomplete`, but rather of Zsh. `zsh-autocomplete` merely enables it and adds some sensible defaults. 

> I do not know 1) how to define `Alt-key` combination for the above keys

The most reliable way to find out what escape code your terminal sends for a particular key or key combo, is as follows:
1. Run `bindkey '^V' describe-key-briefly`.
2. Press <kbd>Ctrl</kbd><kbd>V</kbd>, followed by the key combo you need.

> 2) what the exact menu selection command should I use

You can find a list of those and what they do in [this section of the Zsh manual](https://zsh.sourceforge.io/Doc/Release/Zsh-Modules.html#Menu-selection).

Once you know which escape code and widget to use, you can bind them in the `menuselect` keymap with [Zsh's `bindkey` command](https://zsh.sourceforge.io/Doc/Release/Zsh-Line-Editor.html#index-bindkey) by adding statements like this to your `.zshrc` file:
```zsh
bindkey -M menuselect <keycode> <widget>
```
Restart your shell or terminal for the changes to take effect.

**Reply by marlonrichert** [reply]
@volodymyrprokopyuk Does this solve your problem?

**Reply by volodymyrprokopyuk** [reply]
Hello! Yes, your explanation is exhaustive and complete! Thank you very much for your time and effort! Great job!

**Reply by marlonrichert** [reply]
No problem! Glad to be of help.

PS: If you enjoy using my software, then please consider [sponsoring me](https://github.com/sponsors/marlonrichert). 🙂

**Reply by semi710** [reply]
@volodymyrprokopyuk was you able to remap? if so will you please share the config?

**Reply by volodymyrprokopyuk** [reply]
Unfortunately I did not remapped and stopped using the tool as it was too aggressive in suggestions at that moment.

**Reply by marlonrichert** [reply]
@volodymyrprokopyuk Aggressive in what way? Any suggestions how I could improve this?

**Reply by volodymyrprokopyuk** [reply]
Hi @marlonrichert,

At the time (the time of this issue approximately) when I tried to use the tool I'd like 2 aspects to be improved
* Easily configurable shortcuts through a simple declarations (you suggested to look at Zsh config)
* Easily configurable the scope of suggestions, probably, context sensitive. E. g. only files, only executables, only files if executable is already in place among others. I remember that the zsh-autocomplete provided a huge list of suggestions out-of-the-box

Thank you!

**Reply by marlonrichert** [reply]
> * Easily configurable shortcuts through a simple declarations (you suggested to look at Zsh config)

It is already possible, but I can document it better.

> * Easily configurable the scope of suggestions, probably, context sensitive. E. g. only files, only executables, only files if executable is already in place among others. I remember that the zsh-autocomplete provided a huge list of suggestions out-of-the-box

The plugin already tries to do that. Unfortunately, in various cases, Zsh's completion system doesn't really give fine-grained control over this. However, I'm sure there are areas where I can improve it. Do you have any concrete examples for me where there were too many results? That would help me a lot.

**Reply by volodymyrprokopyuk** [reply]
I've not been using the plugin since the creation of this issue. I'll try to use it again and share my feedback in order to come up with some improvements. Thank you for your attitude and willing to improve the plugin!
